### DISCUSSION TITLE
Tracking Google Pageviews from overlay

### DISCUSSION THREAD

**Discussion by johndwells** [unpoly/unpoly#331] [Q&A] [answered]
We're building an app with a number of pages displayed in an overlay layer. These pages contain links which, when followed, trigger the dismissal of the overlay + a new page being loaded into the root layer.

To integrate with Google Analytics, I have traditionally hooked into the `up:location:changed` event to fire a pageview.

However what I have found is that _dismissing_ the overlay triggers `up:location:changed`, just before another `up:location:changed` is triggered for the new page.

The pageview journey then looks something like this:

/starting-point
 [click to load overlay]
/overlay-page
[click to follow new page]
/starting-point
/new-page

Has anyone encountered this, and figured a way to handle it? Ideally the pageview journey should just be:

/starting-point
/overlay-page
/new-page

:pray:

**Comment by adam12** [comment, marked as answer]
What about [`up:link:follow`](https://unpoly.com/up:link:follow) for event listening instead? Not sure it covers your use case for overlay but it might be a simpler change on your end to test than for me to mock it up in Glitch.

**Reply by johndwells** [reply]
I'm playing with `up:link:follow` now... the one gotcha I see here is that if a layer is dismissed, rather than having a link be followed, then no pageview is triggered. I'm still deciding if I think that's "OK" in this instance... and actually, yes I think that's fine.

Thanks @adam12.

**Reply by johndwells** [reply]
@adam12 After some real world testing, some flaws with this have emerged. The biggest issue is that if I raise a pageview on the `up:link:follow` event, it actually runs _before_ the new page is fetched, rendered and the URL updated. With the way GTM works, this really confuses things. It effectively attempts to log a new pageview, but the current page is still on the stage and the URL is still the current URL...

To mitigate this, I have tried wrapping the `up:link:follow` pageview trigger in a callback function and then calling that function on the `up:location:changed` event, which does improve things.

However the other issue with binding to `up:link:follow` is that back/forward navigating is lost. I have found I can account for this by looking at the `event.reason` parameter passed to `up:location:changed`, and making sure a pageview is triggered here as well.

But it's defo messy...

Alternatively, I have only watching for the `up:location:changed` event, but then wrapping my pageview trigger in a setTimeout function, so that if 2 `up:location:changed` events are fired in quick succession (one for the layer dismissal, one for the new page navigation), then only the new page navigation's setTimeout runs. But the challenge here is network latency - if I set the timeout for 600ms but it takes 1s for the new page to load, then both pageviews will be registered.

Eesh.

**Reply by adam12** [reply]
Are you returning partial responses? If not, maybe an `up.compiler("body")` or `up.compiler("[up-main]")` would be sufficient :joy:

I'm all out of ideas right now. If Henning has time, maybe he has an opinion he can share.

**Reply by johndwells** [reply]
Hmm, no I'm not returning partial responses, so this might be a possibility... But I think this would exclude tracking a simple layer dismissal as a pageview, which technically I _would_ want. I know that sounds picky, but ultimately I'm trying to get pageviews to mimic _user behavior_...

Thanks for your thoughts @adam12! I'll keep mulling this one over, and in the meantime wait to see if Henning has time to consider it...

**Comment by triskweline** [comment]
Two ideas:

- Track a page view on [`up:layer:location:changed`](https://unpoly.com/up:layer:location:changed) instead.
- Track a page view when [rendering a significant piece of content](https://makandracards.com/makandra/41488-using-google-analytics-with-unpoly#section-hard-mode-you-want-more-control-over-what-s-being-tracked), like a main target.

**Reply by johndwells** [reply]
@triskweline Thank you! It looks like `up:layer:location:changed` is the closest to what I'm going for. I will test and report back.

**Reply by johndwells** [reply]
To report back, it does appear that `up:layer:location:changed` works for our needs. A few observations:

1. going from one layer to another (or back to the main) only triggers `up:layer:location:changed` once, and
2. dismissing a modal/overlay layer doesn't trigger `up:layer:location:changed`

The 1st point is exactly what I wanted so this is the big win.

The 2nd point I am happy with and overall makes the most sense in terms of tracking/mimmicking user behavior.

It's worth noting however that I did still wrap my tracking code in a `setTimeout()`, with a duration of `0`, just to give the browser enough time to update the DOM before the pageview was triggered. I did this because of my observations of how GTM was working: as soon as I triggered a new pageview, the information that the GTM container sent also included a scrape/summary of what was on the DOM at the time. But it was happening before the DOM had actually been updated by Unpoly's compilers. Using `setTimeout()` with zero duration moves my pageview code to run after the DOM update.

Here's the final code I'm using:

```javascript
up.on('up:layer:location:changed', (event) => {
  setTimeout((event) => {
    try {
      let pathname = event.location;
      console.log('Track GA pageview: ' + pathname);
      window.dataLayer = window.dataLayer || [];
      window.dataLayer.push({
        'event': 'Pageview',
        'pagePath': pathname
      });
    } catch(e) {
      console.log(e.message);
    }
  }, 0, event);
});
```
